security/user-last-seen.php

```diff
@@ -0,0 +1,148 @@
+	}
+
+	return $user;
+}, 20, 3 );
+
+
+add_filter( 'manage_users_columns', function ( $cols ) {
+	$cols['last_seen'] = __( 'Last seen' );
+	return $cols;
+} );
+
+add_filter( 'manage_users_custom_column', function ( $default, $column_name, $user_id ) {
+	if ( 'last_seen' !== $column_name ) {
+		return $default;
+	}
+
+	$last_seen_timestamp = get_user_meta( $user_id, LAST_SEEN_META_KEY, true );
+
+	if ( ! $last_seen_timestamp ) {
+		return $default;
+	}
+
+	$formatted_date = sprintf(
+		__( '%1$s at %2$s' ),
+		date_i18n( get_option('date_format'), $last_seen_timestamp ),
+		date_i18n( get_option('time_format'), $last_seen_timestamp )
+	);
+
+	if ( ! is_considered_inactive( $user_id ) ) {
+		return sprintf( '<span>%s</span>', esc_html__( $formatted_date ) );
+	}
+
+	$unblock_link = '';
+	if ( current_user_can( 'edit_user', array() ) ) {
+		$url = add_query_arg( array(
+			'action' => 'reset_last_seen',
+			'user_id' => $user_id,
+		) );
+
+		$unblock_link = "<div class='row-actions'><span>User blocked due to inactivity. <a class='reset_last_seen_action' href='" . esc_url( $url ) . "'>" . __( 'Unblock' ) . "</a></span></div>";
+	}
+	return sprintf( '<span class="wp-ui-text-notification">%s</span>' . $unblock_link, esc_html__( $formatted_date ) );
+}, 10, 3 );
+
+add_action( 'user_row_actions', function ( $actions  ) {
+	if( isset($_GET['action'] ) && $_GET['action'] === 'reset_last_seen' ){
+		$user_id = $_GET['user_id'];
+		delete_user_meta( $user_id, LAST_SEEN_META_KEY );
+	}
+
+	return $actions;
+}, 10, 1 );
+
+add_filter( 'views_users', function ( $views ) {
+	global $wpdb;
+
+	if ( ! defined( 'VIP_CONSIDER_USERS_INACTIVE_AFTER_DAYS' ) ) {
+		return $views;
+	}
+
+	$count = $wpdb->get_var( 'SELECT COUNT(meta_key) FROM ' . $wpdb->usermeta . ' WHERE meta_key = "' . LAST_SEEN_META_KEY . '" AND meta_value < ' . get_inactivity_timestamp() );
+
+	$view = __( 'Blocked Users' );
+	if ( $count ) {
+		$class = isset( $_REQUEST[ 'last_seen_filter' ] ) ? 'current' : '';
+		$view = '<a class="' . $class . '" href="users.php?last_seen_filter=blocked">' . $view . '</a>';
+	}
+	$views['blocked_users'] = $view . ' (' . $count . ')';
+
+	return $views;
+} );
+
+add_filter( 'users_list_table_query_args', function ( $args ) {
+	if ( ! defined( 'VIP_CONSIDER_USERS_INACTIVE_AFTER_DAYS' ) ) {
+		return $args;
+	}
+
+	if ( isset( $_REQUEST[ 'last_seen_filter' ] ) ) {
+		$args[ 'meta_key' ] = LAST_SEEN_META_KEY;
+		$args[ 'meta_value' ] = get_inactivity_timestamp();
+		$args[ 'meta_type' ] = 'NUMERIC';
+		$args[ 'meta_compare' ] = '<';
+	}
+
+	return $args;
+} );
+
+function is_considered_inactive( $user_id ) {
+	if ( ! defined( 'VIP_CONSIDER_USERS_INACTIVE_AFTER_DAYS' ) ) {
+		return false;
+	}
+
+	$last_seen_timestamp = get_user_meta( $user_id, LAST_SEEN_META_KEY, true );
+	if ( ! $last_seen_timestamp ) {
+		return false;
+	}
```

Comment:
If a user never logs in to their account, we should probably still deactivate them after the threshold.